### PR TITLE
Mask-RCNN: add PyTorch standard case

### PR DESCRIPTION
**Pull request by yuzhou03**
1. add Mask R-CNN pytorch case
2. COCO2017 dataset verified. 
3. running logs.
[MaskRCNN_logs.zip](https://github.com/FlagOpen/FlagPerf/files/11303010/MaskRCNN_logs.zip)
4. detection results
[MaskRCNN-det-results.zip](https://github.com/FlagOpen/FlagPerf/files/11303093/MaskRCNN-det-results.zip)

5. mAP curve
[Mask-RCNN_mAP_curve.zip](https://github.com/FlagOpen/FlagPerf/files/11303084/Mask-RCNN_mAP_curve.zip)

### CODE REVIEW

File: training/benchmarks/mask_rcnn/pytorch/utils/plot_curve.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
+import os
```
[the remark this answers is not in the document]
Comment (yuzhou03) [reply training/benchmarks/mask_rcnn/pytorch/utils/plot_curve.py:1]: 用于绘制loss和精度曲线

File: training/nvidia/mask_rcnn-pytorch/config/config_A100x2x8.example [hunk 2]
```diff
@@ -0,0 +1,38 @@
+# converting the fp32 parameters to fp16
+fp16 = True # only support true now
+
```
[the remark this answers is not in the document]
Comment (yuzhou03) [reply training/nvidia/mask_rcnn-pytorch/config/config_A100x2x8.example]: removed

File: training/benchmarks/mask_rcnn/pytorch/dataloaders/dataset_voc.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
+import os
```
[the remark this answers is not in the document]
Comment (yuzhou03) [reply training/benchmarks/mask_rcnn/pytorch/dataloaders/dataset_voc.py:1]: removed.

File: training/nvidia/mask_rcnn-pytorch/config/config_A100x2x8.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+fp16 = True
+ddp_type = "apex"
```
[the remark this answers is not in the document]
Comment (yuzhou03) [reply training/nvidia/mask_rcnn-pytorch/config/config_A100x2x8.py]: fixed: updated 2x8 config.

File: training/nvidia/mask_rcnn-pytorch/config/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,8 @@
+torch
```
[the remark this answers is not in the document]
Comment (yuzhou03) [reply training/nvidia/mask_rcnn-pytorch/config/requirements.txt]: removed.

File: training/benchmarks/mask_rcnn/README.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+
+### 模型checkpoint 
+预训练权重下载地址（下载后放入当前文件夹中）
+
+#### Resnet50预训练权重 
+ https://download.pytorch.org/models/resnet50-0676ba61.pth 
+ (注意，下载预训练权重后要重命名， 比如在train.py中读取的是resnet50.pth文件，不是resnet50-0676ba61.pth)
+
+#### Mask R-CNN(Resnet50+FPN)预训练权重
```
[the remark this answers is not in the document]
Comment (yuzhou03) [reply training/benchmarks/mask_rcnn/README.md:52]: 跑我们这个case，这两个pth文件都是需要的。

File: training/run_benchmarks/config/test_conf.py [hunk 7]
```diff
@@ -28,7 +28,7 @@
 
 # The path that flagperf deploy in the cluster.
 # If not set, it will be os.path.dirname(run.py)/../../training/
-FLAGPERF_PATH_HOST = "/home/flagperf/training"
+FLAGPERF_PATH_HOST = "/home/zhouyu/workspace/FlagPerf/training"
```
[the remark this answers is not in the document]
Comment (yuzhou03) [reply training/run_benchmarks/config/test_conf.py]: done

File: training/nvidia/mask_rcnn-pytorch/config/config_A100x1x1.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
+vendor:str = "nvidia"
+data_dir: str = "/mnt/data/maskrcnn/train/"
```
[the remark this answers is not in the document]
Comment (yuzhou03) [reply training/nvidia/mask_rcnn-pytorch/config/config_A100x1x1.py:2]: config里无需填data_dir, 从test_config的cases的value中可以得到。

File: training/nvidia/mask_rcnn-pytorch/config/config_A100x1x1.py [hunk 9]
```diff
@@ -0,0 +1,17 @@
+vendor:str = "nvidia"
+train_batch_size = 16
+eval_batch_size = 16
+
+
+dist_backend = "nccl"
+weight_decay = 1e-4
+adam_beta1 = 0.9
+adam_beta2 = 0.999
```
Comment (shh2000) [inline training/nvidia/mask_rcnn-pytorch/config/config_A100x1x1.py:9]: 这几个adam的参数是做什么用的

File: training/nvidia/mask_rcnn-pytorch/README.md [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+
+### 运行情况
+| 训练资源 | 配置文件        | 运行时长(s) | 目标精度 | 收敛精度(mAP) | 性能（samples/s） |
+| -------- | --------------- | ----------- | -------- | ------------- | ----------------- |
+| 单机1卡  | config_A100x1x1 | 75337.16    | 0.58     | 0.5830        | 28.02             |
+| 单机4卡  | config_A100x1x4 | 24084.01    | 0.58     | 0.5803        | 97.52             |
+| 单机8卡  | config_A100x1x8 | 12154.89    | 0.58     | 0.5803        | 174.20            |
+| 两机8卡  | config_A100x2x8 | 14269.8     | 0.58     | 0.5809        | 271.93            |
+
```
Comment (shh2000) [inline training/nvidia/mask_rcnn-pytorch/README.md:29]: 两机八卡为什么会比单机时间还长？

File: training/benchmarks/mask_rcnn/pytorch/train/trainer.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+            f"pretrain_path:{pretrain_path}, coco_weights_pretrained_path:{coco_weights_pretrained_path}"
+        )
+        self.model = create_model(
+            num_classes=config.num_classes + 1,
+            load_pretrain_weights=True,
+            pretrain_path=pretrain_path,
+            coco_weights_path=coco_weights_pretrained_path)
+        self.model = self.adapter.convert_model(self.model)
+        self.model = self.adapter.model_to_fp16(self.model)
+        # Attention: remember to move model to device before create_optimizer, otherwise, you will get a RuntimeError:
```
Comment (shh2000) [inline training/benchmarks/mask_rcnn/pytorch/train/trainer.py:54]: 这里不需要model_to_ddp吗？是不是和两机八卡时间长但吞吐量高，训练了两遍数据有关

File: training/benchmarks/mask_rcnn/pytorch/coco91_indices.json [hunk 12]
```diff
@@ -0,0 +1,92 @@
+{
```
Comment (shh2000) [inline training/benchmarks/mask_rcnn/pytorch/coco91_indices.json:1]: 这个文件会在dataset调用过程中自动生成，后续使用。不需要提交到stdcase中吧

File: training/benchmarks/mask_rcnn/pytorch/seg_results.json
Comment (shh2000) [inline training/benchmarks/mask_rcnn/pytorch/seg_results.json:1]: 同coco91_indices.json

File: training/benchmarks/mask_rcnn/pytorch/model/__init__.py
Comment (shh2000) [inline training/benchmarks/mask_rcnn/pytorch/model/__init__.py:1]: models/下面这一大堆文件，是否可以替换成torchvision.models.detection.maskrcnn_resnet50_fpn

File: training/benchmarks/mask_rcnn/pytorch/optimizers/factory.py [hunk 13]
```diff
@@ -0,0 +1,9 @@
+import torch
+
+
+def create_optimizer(params, args):
```
Comment (shh2000) [inline training/benchmarks/mask_rcnn/pytorch/optimizers/factory.py:4]: 这个有被使用过吗？我看trainer里面用的是适配器（adapter）里面那个optimizer

File: training/benchmarks/mask_rcnn/pytorch/train/transforms.py
Comment (shh2000) [inline training/benchmarks/mask_rcnn/pytorch/train/transforms.py:1]: 这个transforms使用过吗？我看内容和dataloader下面那个是相同的

File: training/benchmarks/mask_rcnn/pytorch/config/_base.py [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+lr_steps: list = [16, 22]
+# decrease lr by a factor of lr-gamma
+lr_gamma: float = 0.1
+
+max_steps: int = 900000
+seed: int = 41
+
+# Stop training after reaching this accuracy
+target_mAP: float = 0.58
+
```
Comment (shh2000) [inline training/benchmarks/mask_rcnn/pytorch/config/_base.py:42]: 这个用的是bbox map@50吧，是不是应该改用bbox map@all
mmdet详细结果：
{"mode": "val", "epoch": 12, "iter": 7330, "lr": 0.0002, "bbox_mAP": 0.382, "bbox_mAP_50": 0.588, "bbox_mAP_75": 0.414, "bbox_mAP_s": 0.219, "bbox_mAP_m": 0.409, "bbox_mAP_l": 0.495, "bbox_mAP_copypaste": "0.382 0.588 0.414 0.219 0.409 0.495", "segm_mAP": 0.347, "segm_mAP_50": 0.557, "segm_mAP_75": 0.372, "segm_mAP_s": 0.183, "segm_mAP_m": 0.374, "segm_mAP_l": 0.472, "segm_mAP_copypaste": "0.347 0.557 0.372 0.183 0.374 0.472"}
对于multitask的model，是否还应加上对segm mAP的约束
比如这里改成，targetbboxmAP=0.58，targetsegmmAP=0.34，然后在trainer里面读取det_info/seg_info那里改一下。把det_info那个从[1]改成[0]，加上一个seg_info[0]

Comment (shh2000) [reply training/benchmarks/mask_rcnn/pytorch/config/_base.py:42]: mmdet的epoch不用咱们管。他们的epoch/收敛速度普遍比torchvision快一倍，就按照tv的26epoch就行

File: training/benchmarks/mask_rcnn/pytorch/config/_base.py [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+# Sample to begin performing eval.
+eval_iter_start_samples: int = 100
+
+# If set to -1, disable eval, else evaluate every eval_iter_samples during training
+eval_interval_samples: int = 100 * 256 * 1
+
+# Total number of training samples to run.
+max_samples_termination: float = 1388270 * 4 * 30
+
```
Comment (shh2000) [inline training/benchmarks/mask_rcnn/pytorch/config/_base.py:51]: 这个1388270是什么值，后面的30是epoch吗

File: training/benchmarks/mask_rcnn/pytorch/config/_base.py [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+distributed: bool = False
+
+load_pretained: bool = True
+
+# number of classes, background class NOT included
+num_classes: int = 90
+
+# pretrained path
+pretrained_path: str = "checkpoint/resnet50.pth"
+# coco weights pretrained_path
```
Comment (shh2000) [inline training/benchmarks/mask_rcnn/pytorch/config/_base.py:98]: 这个是否需要填None

File: training/benchmarks/mask_rcnn/pytorch/coco91_indices.json (hunk 12 above)
Comment (yuzhou03) [reply training/benchmarks/mask_rcnn/pytorch/coco91_indices.json:1]: removed

File: training/benchmarks/mask_rcnn/pytorch/seg_results.json
Comment (yuzhou03) [reply training/benchmarks/mask_rcnn/pytorch/seg_results.json:1]: removed

File: training/nvidia/mask_rcnn-pytorch/config/config_A100x1x1.py (hunk 9 above)
Comment (yuzhou03) [reply training/nvidia/mask_rcnn-pytorch/config/config_A100x1x1.py]: 这个模型没有用到Adam优化器，removed

File: training/benchmarks/mask_rcnn/pytorch/train/transforms.py
Comment (yuzhou03) [reply training/benchmarks/mask_rcnn/pytorch/train/transforms.py:1]: 重复文件，removed

File: training/nvidia/mask_rcnn-pytorch/README.md (hunk 10 above)
Comment (yuzhou03) [reply training/nvidia/mask_rcnn-pytorch/README.md:29]: 2x8机器资源目前紧缺，等资源可用后，重新跑下多机数据。

File: training/benchmarks/mask_rcnn/pytorch/optimizers/factory.py (hunk 13 above)
Comment (yuzhou03) [reply training/benchmarks/mask_rcnn/pytorch/optimizers/factory.py]: 因为模型的optimizer不应被修改，采用optimizer/factory.py中的create_optimizer，删除trainer_adapter中的create_optimizer方法
建议其他模型也采用这种 方式。

File: training/benchmarks/mask_rcnn/pytorch/train/trainer.py (hunk 11 above)
Comment (yuzhou03) [reply training/benchmarks/mask_rcnn/pytorch/train/trainer.py:57]: trainer.py中，已经添加 model_to_ddp的逻辑。
2x8卡的配置需要待多机资源可用后，重跑一下看结果。

File: training/benchmarks/mask_rcnn/pytorch/config/_base.py (hunk 14 above)
Comment (yuzhou03) [reply training/benchmarks/mask_rcnn/pytorch/config/_base.py:47]: trainer.py的detect_training_status方法，添加了bbox_mAP和 segMAP 两个任务的精度作为收敛条件。

File: training/benchmarks/mask_rcnn/pytorch/config/_base.py (hunk 15 above)
Comment (yuzhou03) [reply training/benchmarks/mask_rcnn/pytorch/config/_base.py:56]: max_samples_termination 已去除，改为用max_epochs来控制train的epoch数。

File: training/benchmarks/mask_rcnn/pytorch/config/_base.py (hunk 16 above)
Comment (yuzhou03) [reply training/benchmarks/mask_rcnn/pytorch/config/_base.py:105]: 不需要，因为不同卡型的nv 配置都只会复用_base.py中的该路径，不会覆盖此配置的值。
